runtime/master/src/main/java/edu/snu/coral/runtime/master/scheduler/SourceLocationAwareSchedulingPolicy.java

```diff
@@ -0,0 +1,178 @@
+   * @return true if the task group is successfully scheduled, false otherwise.
+   */
+  private synchronized boolean attemptSchedule(final ScheduledTaskGroup scheduledTaskGroup,
+                                               final JobStateManager jobStateManager,
+                                               final Set<String> sourceLocations) {
+    final List<ExecutorRepresenter> candidateExecutors =
+        selectExecutorByContainerTypeAndNodeNames(scheduledTaskGroup.getContainerType(), sourceLocations);
+    if (candidateExecutors.size() == 0) {
+      return false;
+    }
+    final int randomIndex = ThreadLocalRandom.current().nextInt(0, candidateExecutors.size());
+    final ExecutorRepresenter selectedExecutor = candidateExecutors.get(randomIndex);
+
+    jobStateManager.onTaskGroupStateChanged(scheduledTaskGroup.getTaskGroupId(), TaskGroupState.State.EXECUTING);
+    selectedExecutor.onTaskGroupScheduled(scheduledTaskGroup);
+    LOG.info("Scheduling {} (source location: {}) to {} (node name: {})", scheduledTaskGroup.getTaskGroupId(),
+        String.join(", ", sourceLocations), selectedExecutor.getExecutorId(),
+        selectedExecutor.getNodeName());
+    return true;
+  }
+
+  @Override
+  public synchronized void onExecutorAdded(final String executorId) {
+    availableExecutors.add(executorId);
+    roundRobinSchedulingPolicy.onExecutorAdded(executorId);
+  }
+
+  @Override
+  public synchronized Set<String> onExecutorRemoved(final String executorId) {
+    availableExecutors.remove(executorId);
+    return roundRobinSchedulingPolicy.onExecutorRemoved(executorId);
+  }
+
+  @Override
+  public void onTaskGroupExecutionComplete(final String executorId, final String taskGroupId) {
+    roundRobinSchedulingPolicy.onTaskGroupExecutionComplete(executorId, taskGroupId);
+  }
+
+  @Override
+  public void onTaskGroupExecutionFailed(final String executorId, final String taskGroupId) {
+    roundRobinSchedulingPolicy.onTaskGroupExecutionFailed(executorId, taskGroupId);
+  }
+
+  /**
+   * @param containerType type of the desired container type
+   * @param nodeNames set of node names
+   * @return list of executors, which resides in one of {@code nodeNames}, has container type of {@code containerType},
+   *         and has an empty slot for execution
+   */
+  private synchronized List<ExecutorRepresenter> selectExecutorByContainerTypeAndNodeNames(
+      final String containerType, final Set<String> nodeNames) {
+    final Map<String, ExecutorRepresenter> executorIdToExecutorRepresenter
+        = containerManager.getExecutorRepresenterMap();
+    final Stream<ExecutorRepresenter> candidates = availableExecutors.stream()
+        .map(executorId -> executorIdToExecutorRepresenter.get(executorId))
+        .filter(executor -> executor.getRunningTaskGroups().size() < executor.getExecutorCapacity())
+        .filter(executor -> nodeNames.contains(executor.getNodeName()));
```

Comment:
It seems that if there are multiple tasks in a task group, each with a different local node (e.g., task A - node X, task B - node Y), then we consider both node X and node Y as local nodes. That is, we take the union of the local nodes of each task. Can you make that explicit in the comment? (e.g., the method-level algorithm description for `scheduleTaskGroup()`)